File: 0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {

public:
    vector<int> postorderTraversal(TreeNode* root) {
        vector<int> ans;
        unordered_set<TreeNode*> visited;
        stack<TreeNode*> s;
        s.push(root);
        while(!s.empty()) {
            TreeNode* curr = s.top();
            if(curr!=NULL) {
                if(visited.find(curr) == visited.end()) {
                    visited.insert(curr);
                    s.push(curr->right);
                    s.push(curr->left);
                } else {
                    ans.push_back(curr->val);
                    s.pop();
                }
            } else {
                s.pop();
            }
        }
        return ans;
    }
};

File: 0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> postorderTraversal(TreeNode root) {
      List<Integer> ans = new ArrayList<>();
        return travel(root, ans);
    }

    private List<Integer> travel(TreeNode root, List<Integer> ans) {
        if (root == null)
            return ans;
        travel(root.left, ans);
        travel(root.right, ans);
        ans.add(root.val);
        return ans;
    }
}